[DL-5849] Transparent impersonations

Windvis opened the pull request:
After some experimentation and feedback we came to the conclusion that the previous implementation isn't transparent enough. Each micro service that accesses session data would need to know about the impersonation which isn't ideal. 

Instead of adding a `muAccount:impersonates` predicate, we override the actual relevant data on the sessios, so the impersonation is "transparent" throughout the stack. This requires app specific knowledge though, so it's not generic at the moment. (The PR assumes an LBLOD app that uses the acmidm-login-service or mock-login-service). We could try to make things configurable so other apps can use it as well.

**TODO**:
- [x] Fix the "switch impersonation", at the moment it only works for the first impersonation
- [x] Update the mu-auth config example so it works for the new setup
- [ ] double check if the new predicates are good

**Nice to have:**
- [ ] ~~configuration options so it can be used by non-lblod projects and we can upstream the changes to the base repo (and move it to rpio or mu-semtech).~~ Out of scope for now.

**Current issues when testing in the** [loket PR](https://github.com/lblod/app-digitaal-loket/pull/545):
- ~~stopping the impersonation doesn't work (mu-auth config issue)~~
- ~~switching sessions doesn't work (issue here, and mu-auth config issue)~~
- ~~Loket needs the "group" of the original account to display it in the header, and roles so we can check if the user is an admin (although we can technically assume that if an impersonation is active, since only admins can do that). We don't return these yet. If we do, it's an extra non-generic thing, since not all apps have session groups so it makes it harder to upstream the changes.~~

Windvis committed:
Transparent impersonations

Instead of adding a muAccount:impersonates predicate, we override the actual data, so the impersonation is "transparent" throughout the stack.

Windvis committed:
Fix the setImpersonatedSession query

The switch flow wasn't working properly. When switching to a different account impersonation the roles and groups would be copied to the `ext:originalX` predicates because the FILTER NOT EXISTS part of the query was wrong. If you then stopped impersonating the roles would still be attached to the admin account.

Windvis committed:
Update the readme

The new version requires a different mu-auth setup and the API response is changed a bit.

Windvis commented on lib/session.js:
```diff
@@ -2,45 +2,134 @@ import { query, update, sparqlEscapeUri } from 'mu';
+      BIND(${sparqlEscapeUri(sessionUri)} AS ?uri)
+      ?uri mu:uuid ?id ;
+        muSession:account ?impersonatedResource ;
+        muExt:originalResource ?originalResource ;
```
Since we're in forked space now we can name this more specifically if we want: `muExt:originalAccount`?

Riadabd replied:
I agree with `muExt:originalAccount` as it feels "clearer" in a sense, especially since we use `muSession:account` to point to the actual session. But, we can wait for Felix to see what to go for.

cecemel replied:
not really an opinion ATM. This service is going to spread over multiple stacks. how much more or less work will it be if we also decide to work towards a more generic version of this service?

Windvis replied:
I think it wouldn't change much since the queries would need to be configurable anyways.

Windvis commented on app.js:
```diff
@@ -25,10 +28,19 @@ app.get('/impersonations/current', async function (req, res, next) {
         links: `/resources/${impersonatedResourceId}`,
         data: { type: 'resources', id: impersonatedResourceId },
+      },
+      'original-resource': {
```
Same here, we could name this `original-account`.

cecemel replied:
maybe to be in consistent with the fact that we are tailoring with acm/abb setup.

Windvis committed:
fixup readme

Windvis committed:
Use "account" instead of "resource"

This is more specific in the context of the LBLOD apps and makes the code easier to follow.

Windvis committed:
Update the readme

Windvis committed:
readme fixup
